arwen/host/asyncCall.go

```diff
@@ -0,0 +1,150 @@
+	callbackCallInput, err := host.createCallbackContractCallInput(destinationVMOutput)
+	if err != nil {
+		return err
+	}
+
+	_, err = host.ExecuteOnDestContext(callbackCallInput)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (host *vmHost) canExecuteSynchronously() bool {
+	runtime := host.Runtime()
+	blockchain := host.Blockchain()
+	asyncCallInfo := runtime.GetAsyncCallInfo()
+	dest := asyncCallInfo.Destination
+	calledSCCode, err := blockchain.GetCode(dest)
+
+	return len(calledSCCode) != 0 && err == nil
+}
+
+func (host *vmHost) createDestinationContractCallInput() (*vmcommon.ContractCallInput, error) {
+	runtime := host.Runtime()
+	sender := runtime.GetSCAddress()
+	asyncCallInfo := runtime.GetAsyncCallInfo()
+
+	argParser := runtime.ArgParser()
+	err := argParser.ParseData(string(asyncCallInfo.Data))
+	if err != nil {
+		return nil, err
+	}
+
+	function, err := argParser.GetFunction()
+	if err != nil {
+		return nil, err
+	}
+
+	arguments, err := argParser.GetArguments()
+	if err != nil {
+		return nil, err
+	}
+
+	gasLimit := asyncCallInfo.GasLimit - host.Metering().GasSchedule().ElrondAPICost.AsyncCallStep
+	if gasLimit <= 0 {
+		return nil, arwen.ErrNotEnoughGas
+	}
+
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  sender,
+			Arguments:   arguments,
+			CallValue:   big.NewInt(0).SetBytes(asyncCallInfo.ValueBytes),
+			CallType:    vmcommon.AsynchronousCall,
+			GasPrice:    0,
+			GasProvided: gasLimit,
+		},
+		RecipientAddr: asyncCallInfo.Destination,
+		Function:      function,
+	}
+
+	return contractCallInput, nil
+}
+
+func (host *vmHost) createCallbackContractCallInput(destinationVMOutput *vmcommon.VMOutput) (*vmcommon.ContractCallInput, error) {
+	metering := host.Metering()
+	runtime := host.Runtime()
+
+	arguments := destinationVMOutput.ReturnData
+	gasLimit := destinationVMOutput.GasRemaining
+	function := "callBack"
+
+	dataLength := host.computeDataLengthFromArguments(function, arguments)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.AsyncCallStep
+	gasToUse += metering.GasSchedule().BaseOperationCost.DataCopyPerByte * uint64(dataLength)
+	gasLimit -= gasToUse
+	if gasLimit <= 0 {
+		return nil, arwen.ErrNotEnoughGas
+	}
+
+	sender := runtime.GetAsyncCallInfo().Destination
+	dest := runtime.GetSCAddress()
+
+	// Return to the sender SC, calling its callback() method.
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  sender,
+			Arguments:   arguments,
+			CallValue:   big.NewInt(0),
+			CallType:    vmcommon.AsynchronousCallBack,
+			GasPrice:    0,
+			GasProvided: gasLimit,
+		},
+		RecipientAddr: dest,
+		Function:      function,
+	}
+
+	return contractCallInput, nil
+}
+
+func (host *vmHost) computeDataLengthFromArguments(function string, arguments [][]byte) int {
+	// Calculate what length would the Data field have, were it of the
+	// form "callback@arg1@arg4...
+	dataLength := len(function) + 1
```

Comment:
dataLength = len(function) + SUM(len(argument)) + len(arguments)

Reply:
Right. In the current form, `computeDataLengthFromArguments()` ignores each empty argument. We need to change how `VMOutput.ReturnData` is handled, because empty entries are important, e.g. when interpreting `ReturnData` as arguments for `callBack()`. Currently, we skip these entries (see `Finish` functions) and this should be changed. Added a TODO and noted the issue.